framework/db/ColumnSchema.php

```diff
@@ -114,12 +115,18 @@ public function dbTypecast($value)
      */
     protected function typecast($value)
     {
-        if ($value === '' && $this->type !== Schema::TYPE_TEXT && $this->type !== Schema::TYPE_STRING && $this->type !== Schema::TYPE_BINARY && $this->type !== Schema::TYPE_CHAR) {
+        if ($value === '' && !in_array($this->type, [Schema::TYPE_TEXT, Schema::TYPE_STRING, Schema::TYPE_BINARY, Schema::TYPE_CHAR], true)) {
             return null;
         }
-        if ($value === null || gettype($value) === $this->phpType || $value instanceof Expression || $value instanceof Query) {
+        if ($value === null || gettype($value) === $this->phpType || $value instanceof ExpressionInterface || $value instanceof Query) {
             return $value;
         }
+        if (is_array($value) && count($value) === 2 && isset($value[1]) && in_array($value[1], $this->getPdoParamTypes(), true)) {
+            return new PdoValue($value[0], $value[1]);
+            // Backward compatibility thing, but too unreliable. Array [1,2] will pass into this condition and transform
+            // into PdoValue class. However, arrays did not come this method earlier so should not be a big deal. TODO: consider.
```

Comment:
> However, arrays did not come this method earlier

Are you sure about that? Was it impossible to supply an array here?

Reply:
It was used in `dbTypecast()` and `phpTypecast()` methods only.

Here are the complete list of usages:

**dbTypecast()**

 - `QueryBuilder::insert()` prefixed with `!is_array()`
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/QueryBuilder.php#L195-L197

- `QueryBuilder::batchInsert()` wrapped in `!is_array()` condition
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/QueryBuilder.php#L279-L281

- `QueryBuilder::update()` prefixed with `!is_array()`
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/QueryBuilder.php#L344-L346

- MySQL `QueryBuilder::insert()` prefixed with `!is_array()`
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/mysql/QueryBuilder.php#L259-L251

- Oracle `QueryBuilder::insert()` prefixed with `!is_array()`
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/oci/QueryBuilder.php#L212-L215

- Oracle `QueryBuilder::batchInsert()` wrapped in `!is_array()` condition
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/oci/QueryBuilder.php#L268

- PgSQL `QueryBuilder::batchInsert()` wrapped in `!is_array()` condition
https://github.com/yiisoft/yii2/blob/master/framework/db/pgsql/QueryBuilder.php#L305-L307

- SQLite `QueryBuilder::batchInsert()` wrapped in `!is_array()` condition
https://github.com/yiisoft/yii2/blob/master/framework/db/sqlite/QueryBuilder.php#L101-L103

**phpTypecast()**

- `ActiveRecord::populateRecord()` – potentially dangerous
https://github.com/yiisoft/yii2/blob/5a8c3d537ba14dbff02b2adee9072d381b9eb6b7/framework/db/ActiveRecord.php#L418-L427

- `ActiveRecord::insertInternal()` – have you ever seen primary key represented as array of arrays?
https://github.com/yiisoft/yii2/blob/5a8c3d537ba14dbff02b2adee9072d381b9eb6b7/framework/db/ActiveRecord.php#L515-L517

- Also used to typecast default values: [1](https://github.com/yiisoft/yii2/blob/1501c659aca7c50fa5e4f6fbe5ca549acc946c86/framework/db/cubrid/Schema.php#L335), [2](https://github.com/yiisoft/yii2/blob/4e176ca3259083a8c441c70182bc10e96cbfec12/framework/db/mssql/Schema.php#L386), [3](https://github.com/yiisoft/yii2/blob/f9fd4da8de6c0272cd8956e94da1656b41f1416c/framework/db/mysql/Schema.php#L303), [4](https://github.com/yiisoft/yii2/blob/185209957e11ba6a427c1aa418dfcba7734e8b7b/framework/db/oci/Schema.php#L416), [5](https://github.com/yiisoft/yii2/blob/284dae6a55e44590d495831b9761f1ca50a8bc76/framework/db/pgsql/Schema.php#L565), [6](https://github.com/yiisoft/yii2/blob/2d672b67223b8d930f0f444b113d0a13fbd551c0/framework/db/sqlite/Schema.php#L364), but should not be harmful.

Considering that tests are not broken, I tend to say that the change is not a serious breaking.

Reply:
Yes. There could be composite PKs. These contain multiple columns.

Reply:
No, composite PK is array of strings: `['id', 'version']`, not array of arrays `[['id'], ['version']]`

Reply:
Then it should be fine.